File: strings/check-if-two-string-arrays-are-equivalent/check-if-two-string-arrays-are-equivalent.js
function arrayStringsAreEqual(word1, word2) {
  let i = 0;
  let j = 0;

  for (let a = 0, b = 0; i < word1.length && j < word2.length; ) {
    const currString1 = word1[i];
    const currChar1 = currString1[a++];

    const currString2 = word2[j];
    const currChar2 = currString2[b++];

    if (currChar1 !== currChar2) {
      return false;
    }

    if (a === currString1.length) {
      a = 0;
      i++;
    }

    if (b === currString2.length) {
      b = 0;
      j++;
    }
  }

  return i === word1.length && j === word2.length;
}

File: strings/check-if-two-string-arrays-are-equivalent/check-if-two-string-arrays-are-equivalent.go
func arrayStringsAreEqual(word1 []string, word2 []string) bool {
	i := 0
	j := 0

	for a, b := 0, 0; i < len(word1) && j < len(word2); {
		currString1 := word1[i]
		currChar1 := currString1[a]
		a++

		currString2 := word2[j]
		currChar2 := currString2[b]
		b++

		if currChar1 != currChar2 {
			return false
		}

		if a == len(currString1) {
			a = 0
			i++
		}

		if b == len(currString2) {
			b = 0
			j++
		}
	}

	return i == len(word1) && j == len(word2)
}
